src/util/priority_queue.c

```diff
@@ -0,0 +1,159 @@
+/*
+* Copyright 2018-2020 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "./priority_queue.h"
+#include <stdlib.h>
+#include <stdio.h>
+
+/**
+ * @brief  Initialize a Queue node, and its undelying CacheData.
+ * @param  node: Node pointer.
+ * @param  hashKey: Node's cache key (for CacheData).
+ * @param  resultSet: Node's result value (for CacheData).
+ * @retval Initialized Queue node (pointer)
+ */
+static QueueItem *_PriorityQueue_InitQueueItem(QueueItem *node, void *cacheValue, size_t dataSize,
+											   QueueDataFreeFunc freeCB) {
+	// If node was in use before, call its destructor.
+	if(node->isDirty) {
+		void *data = (void *)node->data;
+		if(freeCB) freeCB(data);
+	}
+	// Set new value and mark as used.
+	memcpy(&node->data, cacheValue, dataSize);
+	node->isDirty = true;
+
+	return node;
+}
+
+/**
+ * @brief  Returns the size of node given a size of data value.
+ * @param  dataSize:
+ * @retval The size of the queue node.
+ */
+static inline size_t _PriorityQueue_SizeOfQueueItem(size_t dataSize) {
+	return dataSize + sizeof(LinkedListNode) + sizeof(bool);
+}
+
+PriorityQueue *PriorityQueue_Create(size_t capacity, size_t dataSize, QueueDataFreeFunc freeCB) {
+	// Memory allocations.
+	PriorityQueue *queue = rm_malloc(sizeof(PriorityQueue));
+	queue->buffer = rm_calloc(capacity, _PriorityQueue_SizeOfQueueItem(dataSize));
+	queue->freeList = array_new(QueueItem *, capacity);
+	queue->dataSize = dataSize;
+	queue->freeCB = freeCB;
+	// Maximal capacity.
+	queue->capacity = capacity;
+	// Initial size is 0.
+	queue->size = 0;
+	// First empty space is the first place in the queue array.
+	queue->emptySpace = queue->buffer;
+	// Queue hasn't reached full capacity, a linear insertion is possible.
+	queue->stopLinearInsertion = false;
+	LinkedList_Init(&queue->linked_list);
+	return queue;
+}
+
+inline bool PriorityQueue_IsFull(const PriorityQueue *queue) {
+	// Maximal capacity reached.
+	return queue->capacity == queue->size;
+}
+
+inline bool PriorityQueue_IsEmpty(const PriorityQueue *queue) {
+	return queue->size == 0;
+}
+
+inline void *PriorityQueue_Dequeue(PriorityQueue *queue) {
+	// For empty queue return null.
+	if(PriorityQueue_IsEmpty(queue)) return NULL;
+	LinkedListNode *head = queue->linked_list.head;
+	LinkedList_RemoveNode(&queue->linked_list, head);
+	QueueItem *queue_item = (QueueItem *) head;
+	array_append(queue->freeList, queue_item);
```

Comment:
While the array will not realloc thanks to its fixed size, let's follow convention:
`queue->freeList = array_append(queue->freeList, queue_item);`